File: packages/septr-go/strip.go
package septr

func stripSensitiveData(obj interface{}, customFields ...[]string) (interface{}, []DetectionEvent) {
	var detections []DetectionEvent
	var clean func(interface{}) interface{}
	clean = func(value interface{}) interface{} {
		if value == nil {
			return nil
		}
		switch v := value.(type) {
		case string:
			specific := detectSecrets(v)
			entropy := DetectHighEntropySecrets(v)
			if len(specific) > 0 || len(entropy) > 0 {
				detections = append(detections, specific...)
				detections = append(detections, entropy...)
				// Redact when at least one specific detection is redactable.
				// Advisory-only patterns (e.g. public API keys) skip redaction
				// but must not suppress redaction of real secrets alongside them.
				foundRedactable := false
				for _, s := range specific {
					if s.Redactable == nil || *s.Redactable {
						foundRedactable = true
						break
					}
				}
				if foundRedactable {
					return "[REDACTED]"
				}
			}
			return v
		case []interface{}:
			result := make([]interface{}, len(v))
			for i, item := range v {
				result[i] = clean(item)
			}
			return result
		case map[string]interface{}:
			result := make(map[string]interface{})
			for key, val := range v {
				if shouldStripKey(key, customFields...) {
					detections = append(detections, DetectionEvent{
						Type: "data_strip", Severity: "medium",
						PatternID: "strip_field",
						Description: "Field `" + key + "` stripped from response",
						StatusCode: 200, Timestamp: nowMs(),
					})
					result[key] = "[REDACTED]"
				} else {
					result[key] = clean(val)
				}
			}
			return result
		default:
			return value
		}
	}
	return clean(obj), detections
}

File: packages/septr-python/septr/core/strip.py
import time
from typing import Any, Optional
from .secrets import detect_secrets, detect_high_entropy_secrets, should_strip_key, DetectionEvent


def strip_sensitive_data(
    obj: Any,
    custom_fields: Optional[list[str]] = None,
) -> tuple[Any, list[DetectionEvent]]:
    detections: list[DetectionEvent] = []

    def clean(value: Any) -> Any:
        nonlocal detections
        if value is None:
            return value
        if isinstance(value, str):
            specific = detect_secrets(value)
            entropy = detect_high_entropy_secrets(value)
            if specific or entropy:
                detections.extend(specific)
                detections.extend(entropy)
                # Redact when at least one specific detection is redactable.
                # Advisory-only patterns (e.g. public API keys) skip redaction
                # but must not suppress redaction of real secrets alongside them.
                if any(d.redactable is not False for d in specific):
                    return "[REDACTED]"
            return value
        if isinstance(value, list):
            return [clean(item) for item in value]
        if isinstance(value, dict):
            cleaned: dict[str, Any] = {}
            for key, val in value.items():
                if should_strip_key(key, custom_fields):
                    detections.append(DetectionEvent(
                        type="data_strip", severity="medium",
                        patternId="strip_field",
                        description=f"Field `{key}` stripped from response",
                        statusCode=200,
                        timestamp=time.time() * 1000,
                    ))
                    cleaned[key] = "[REDACTED]"
                else:
                    cleaned[key] = clean(val)
            return cleaned
        return value

    return clean(obj), detections
